### PR TITLE
[FSDP] Add limiter using CUDA events

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#83666 [FSDP] Add limiter using CUDA events**
* #83665 [FSDP] Generalize prefetching; lower unshard/reshard to handle
* #83664 [FSDP] ufmt `flat_param.py`, `flatten_params_wrapper.py`
* #83663 [FSDP] Retire `self.device_id`; clean up ctor

This PR tackles the high GPU reserved memory issue for FSDP.

Currently:
- This adds an argument `all_gather_issue_limit: Optional[int]` to the FSDP constructor, where `None` disables the limiter and a positive integer enables the limiter.
- If enabled, this limiter is only meaningful for `FULL_SHARD` and not for `SHARD_GRAD_OP` and `NO_SHARD` (since (1) we track free events, not all-gather events and (2) for the non-`FULL_SHARD` strategies, the reserved memory will inevitably be used).
- Given this, ideally each sharding strategy can have its own attributes, and we can move this `all_gather_issue_limit` to only be an attribute for `FULL_SHARD`. This idea also applies to `HYBRID_SHARD` since one option then is to pass the second process group as an attribute there.
- I want to discuss this since this does not seem backward compatible. I am not sure that with [enums](https://stackoverflow.com/questions/12680080/python-enums-with-attributes), we can have different attributes per enum.

#### [Fairscale FSDP Approach 1](https://github.com/facebookresearch/fairscale/pull/972/)
- [Record pre-forward order](https://github.com/facebookresearch/fairscale/blob/191553190d73a5ef4a48687c889d4b1d94532135/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1392-L1394)
- [Use pre-forward prefetching](https://github.com/facebookresearch/fairscale/blob/191553190d73a5ef4a48687c889d4b1d94532135/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1406-L1412) (pre-forward order index + 1)
- [Use pre-backward prefetching](https://github.com/facebookresearch/fairscale/blob/191553190d73a5ef4a48687c889d4b1d94532135/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1502-L1507) (pre-forward order index - 1)
- Prefetch before freeing the padded unsharded flattened parameter
- [Synchronize the current stream after freeing the padded unsharded flattened parameter](https://github.com/facebookresearch/fairscale/blob/191553190d73a5ef4a48687c889d4b1d94532135/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L2073) (regardless of prefetch)

#### [Fairscale FSDP Approach 2](https://github.com/facebookresearch/fairscale/pull/1052)
- [Record post-forward order](https://github.com/facebookresearch/fairscale/blob/7d46cba0ac2bc2d69922d75a454c08edf07bb6ce/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1430-L1431)
- [Record pre-backward order](https://github.com/facebookresearch/fairscale/blob/7d46cba0ac2bc2d69922d75a454c08edf07bb6ce/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1521-L1522)
- [Use post-forward prefetching](https://github.com/facebookresearch/fairscale/blob/7d46cba0ac2bc2d69922d75a454c08edf07bb6ce/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1434) (post-forward index + 1)
- [Use post-backward prefetching](https://github.com/facebookresearch/fairscale/blob/7d46cba0ac2bc2d69922d75a454c08edf07bb6ce/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1675) (pre-backward index + 1)
- [Prefetch after freeing the padded unsharded flattened parameter](https://github.com/facebookresearch/fairscale/blob/7d46cba0ac2bc2d69922d75a454c08edf07bb6ce/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L2106)
- [Synchronize the current stream after freeing the padded unsharded flattened parameter](https://github.com/facebookresearch/fairscale/blob/7d46cba0ac2bc2d69922d75a454c08edf07bb6ce/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L2132) (regardless of prefetch)

#### PT-D FSDP Approach
- In `_reshard()`, record a CUDA event after freeing the padded unsharded flattened parameter
- In `_unshard()`, before actually unsharding, check if the number of saved free events exceeds a max number and if so, synchronize the earliest event, blocking the CPU thread until that event completes

#### T5 (500M) 2 Nodes 16 A100 GPUs 256 Batch Size

<details>
  <summary> `all_gather_issue_limit=None` </summary>
  
Screen Shot 2022-08-16 at 4 51 25 PM

Peak GPU reserved memory: 6784 MB = 6.784 GB
Time / batch: 3.4 s

</details>

<details>
  <summary> `all_gather_issue_limit=2` </summary>
  
Screen Shot 2022-08-16 at 4 51 14 PM

Peak GPU reserved memory: 5846 MB = 5.846 GB
Time / batch: 3.4 s

</details>

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -889,6 +889,12 @@ class FullyShardedDataParallel(nn.Module):
             This can also help load checkpoints taken by ``state_dict`` and to be loaded by
             ``load_state_dict`` in a memory efficient way. See documentation for
             :class:`FullStateDictConfig` for an example of this. (Default: ``False``)
+        all_gather_issue_limit (Optional[int]): If this is not ``None`` and
+            the sharding strategy is ``FULL_SHARD``, this represents the
+            maximum number of actively issued all-gathers. When this limit is
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:894]: "maximum number of in-flight all-gathers"?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -889,6 +889,12 @@ class FullyShardedDataParallel(nn.Module):
...
+            reached, FSDP blocks the CPU thread to ensure that some FSDP
+            parameters are freed before issuing further all-gathers. (Default:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:896]: mention what this can help (gpu memory) and how to configure it?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -960,6 +969,7 @@ def __init__(
         self.cpu_offload = cpu_offload or CPUOffload()
         self.backward_prefetch = backward_prefetch
         self.forward_prefetch = forward_prefetch
+        self.all_gather_issue_limit = all_gather_issue_limit
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:972]: this mostly should apply to the root FSDP unit, i.e. all FSDP units should share the same limit? what happens with manual wrapping if users specify it differently?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -1458,6 +1469,14 @@ def _unshard(
         Postcondition: Each handle's ``FlatParameter`` 's data is the padded
         unsharded flattened parameter on the compute device.
         """
+        if (
+            self.all_gather_issue_limit is not None
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1473]: might be able to make this check less verbose by defaulting the limit to math.maxint or something.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -1736,6 +1759,7 @@ def _lazy_init(self) -> None:
                 fsdp_module._is_root = False
                 fsdp_module._streams = self._streams
                 fsdp_module._exec_order_data = self._exec_order_data
+                fsdp_module._free_events = self._free_events
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1762]: should we also share _all_gather_issue_limit?

**Review by rohan-varma — COMMENTED** [review]
Looks good overall, but I guess the API is still pending discussions? A couple of suggestions:

1. I see the initial experimentation on T5, can we find the set of representative workloads where we saw high GPU memory issue and evaluate impact of this diff on those as well? 
2. Maybe full_param_size is a better metric to ratelimit on, because # of all_gathers will have to be tuned if we shard differently / more / less granular? 
3. Does this also ratelimit the backward?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:896]: I am still working on getting a complete understanding. I will update the docstring once I do!

**Comment by awgu** [comment]
Closing this since it is stale: See https://github.com/pytorch/pytorch/pull/83917 instead.
